docs/elements.md

```diff
@@ -999,7 +999,8 @@ def generate(data):
 | `placeholder`             | text                | None     | Hint displayed inside the input box describing the expected type of input.                                                                                                |
 | `size`                    | integer             | 35       | Width of the input box.                                                                                                                                                   |
 | `show-help-text`          | boolean             | true     | Show the question mark at the end of the input displaying required input parameters.                                                                                      |
-| `multiline`               | boolean             | false    | Whether or not to to allow the input to include line breaks.                                                                                                              |
+| `multiline`               | boolean             | false    | Whether or not to allow the input to include line breaks.                                                                                                                 |
+| `escape-unicode`          | boolean             | true     | Whether or not to escape unprintable unicode characters. Default is `false` if `multiline` is enabled.                                                                    |
```

Comment:
Could we default this to `true` even for multiline input or is there a problem with CR/LF characters?

It [sounds like](https://stackoverflow.com/a/14217315) `textarea` elements will always send us `CR+LF` line endings, so I think it should work to just not escape the `CR+LF` pair on the end of each line, but to escape everything else within the line (including any singleton `CR` or `LF` characters that aren't part of a `CR+LF` pair).

Reply:
Personally I'm in favour of normalizing both correct answer (in prepare) and submitted answer (in parse) to `LF`, mostly because I believe many instructors will likely use `\n` in their setting of the correct answer, and trying to explain this difference may be something we want to avoid. In essence, replace `/\r?\n/` (or, if we're feeling generous, `/\\s*\n/`) with `\n` in parse. In either case, I agree that the line break should not be escaped, whether or not we keep the `CR`.

Reply:
> Personally I'm in favour of normalizing both correct answer (in prepare) and submitted answer (in parse) to `LF`, mostly because I believe many instructors will likely use `\n` in their setting of the correct answer, and trying to explain this difference may be something we want to avoid. In essence, replace `/\r?\n/` (or, if we're feeling generous, `/\\s*\n/`) with `\n` in parse.

I'm mostly in agreement with this, and the current behavior in this PR is to normalize the submitted answer to `LF` (https://github.com/PrairieLearn/PrairieLearn/pull/9438/files#diff-e3c5943c659a94abbfdbf1f52defa85320b1271a71523abfb1a4b54f17131a9aR254). It's straightforward enough to add this behavior to the correct answer used in grade as well (doing this in grade rather than parse is consistent with how other input transformations are handled).

> Could we default this to true even for multiline input or is there a problem with CR/LF characters?

I'm actually in favor of keeping the current behavior of an all-or-nothing escaping of the `Cc` (unicode control) characters (so by default, no escaping for multiline input). The reason for this is that the point of the multiline input is to faithfully display the input string with control characters (so newlines skip to the next line, tabs indent, etc.). In fact, I think newlines are the most common examples of the control characters, and it's much less confusing than with the one-line string input if these are present (looking back, I think one of the main reasons for the previous escaping behavior was probably just to avoid displaying the student output on multiple lines).

To be clear, outside of the `Cc` and `Cf` character codes, other weird unicode stuff is handled by `normalize-to-ascii` option, so changing the escaping of control characters won't introduce weird issues outside of changing the display.

Reply:
> I'm mostly in agreement with this, and the current behavior in this PR is to normalize the submitted answer to `LF` (https://github.com/PrairieLearn/PrairieLearn/pull/9438/files#diff-e3c5943c659a94abbfdbf1f52defa85320b1271a71523abfb1a4b54f17131a9aR254). It's straightforward enough to add this behavior to the correct answer used in grade as well (doing this in grade rather than parse is consistent with how other input transformations are handled).

Sounds great!

Should we also normalize singleton `\r` characters (not followed by an `\n`) to `\n`?

> I'm actually in favor of keeping the current behavior of an all-or-nothing escaping of the `Cc` (unicode control) characters (so by default, no escaping for multiline input). The reason for this is that the point of the multiline input is to faithfully display the input string with control characters (so newlines skip to the next line, tabs indent, etc.). In fact, I think newlines are the most common examples of the control characters, and it's much less confusing than with the one-line string input if these are present (looking back, I think one of the main reasons for the previous escaping behavior was probably just to avoid displaying the student output on multiple lines).
> 
> To be clear, outside of the `Cc` and `Cf` character codes, other weird unicode stuff is handled by `normalize-to-ascii` option, so changing the escaping of control characters won't introduce weird issues outside of changing the display.

I think there are two main problems we face:
1. How do we help instructors write questions that don't suffer from unexpected grading behavior caused by students inputting unexpected characters?
2. How do we help instructors (and students) see exactly what was input, especially in cases where the submission was graded as incorrect but it looks visually correct?

For problem (1) I think that normalizing CR+LF and CR to LF is probably sufficient, as you say. For (2), the `normalize-to-ascii` option handles this, but it defaults to false and so is of no use post hoc and there are also legitimate reasons for not using it. To solve (2), some ideas are:
1. Escape unicode Cc and Cf sets in multiline mode, except for TAB and LF (assuming we normalize CR to LF). Of course this would only be a partial solution.
2. Add some other way for instructors (and students) to see exactly what was input. For example, this could be a "?" button next to the input in the submitted answer that shows a popover with input displayed with every non-printable-ascii (that is, everything outside of 0x20 to 0x7E) rendered as `<U+....>`.

What do you think? We could defer this to a later PR, but I do think it becomes more important with multiline input because students are more likely to be copy/pasting blocks of text that may well contain characters that the question author did not expect.

Reply:
I agree with you that solution (2) is better than (1), both because it builds consistent expectations across elements and also because it allows us to escape all unicode characters, not just Cc and Cf. As you say, option (1) adds complexity without fully solving the problem.

Implementing option (2) in a followup PR would be great, but I'm also on board with squeezing it into this PR if it's small.

Reply:
Yes, the `pl-symbolic-input` doesn't have any special behavior for unicode escaping in its display of the raw answer. I agree standardizing the display in this case would be valuable. This will require a little more thought (and namely, a second PR), but I think we could do this with pretty low overhead overall by sharing the logic. As long as what's going on here currently is semi-reasonable (I'd argue it's probably an improvement over the previous behavior), I should be able to handle this in a short follow-up.

Reply:
Made an issue for this follow-up: https://github.com/PrairieLearn/PrairieLearn/issues/9717

Given the scope, it might not be a super fast follow-up (I would be hesitant to deploy unifying logic like this before summer), but discussed during the meeting and major changes to the escaping behavior are a little orthogonal to the point of this PR.

I think the bar for review on the present PR should be just be adding the multiline behavior and making it no-more-confusing than the present escaping, while compatible with the later planned changes. For this, I will remove the element attribute for escaping on the display (since ultimately, we want to display with a popover that shows all the information instead of adding an attribute), and add a popover with a barebones escaping of the string inputted by the student (using the same visual language as the raw display on symbolic input). We can defer larger changes / discussion to the linked issue.

Reply:
Initial changes in this direction here, still have to work out the kinks: https://github.com/PrairieLearn/PrairieLearn/pull/9438/commits/1e734b640bf5156b22cea944515714533477b968

Reply:
I like your proposed "no-more-confusing-than-present" strategy.